wallet/wallet.c

```diff
@@ -232,3 +235,109 @@ s64 wallet_get_newindex(struct lightningd *ld)
 	db_set_intvar(ld->wallet->db, "bip32_max_index", newidx);
 	return newidx;
 }
+
+bool wallet_shachain_init(struct wallet *wallet, struct wallet_shachain *chain)
+{
+	/* Create shachain */
+	shachain_init(&chain->chain);
+	if (!db_exec(__func__, wallet->db,
+		     "INSERT INTO shachains (min_index, num_valid) VALUES (0,0);")) {
+		return false;
+	}
+	chain->id = sqlite3_last_insert_rowid(wallet->db->sql);
+	return true;
+}
+
+/* TODO(cdecker) Stolen from shachain, move to some appropriate location */
+static unsigned int count_trailing_zeroes(shachain_index_t index)
+{
+#if HAVE_BUILTIN_CTZLL
+	return index ? (unsigned int)__builtin_ctzll(index) : SHACHAIN_BITS;
+#else
+	unsigned int i;
+
+	for (i = 0; i < SHACHAIN_BITS; i++) {
+		if (index & (1ULL << i))
+			break;
+	}
+	return i;
+#endif
+}
+
+bool wallet_shachain_add_hash(struct wallet *wallet,
+			      struct wallet_shachain *chain,
+			      shachain_index_t index,
+			      const struct sha256 *hash)
+{
+	tal_t *tmpctx = tal_tmpctx(wallet);
+	u32 pos = count_trailing_zeroes(index);
+	assert(index < SQLITE_MAX_UINT);
+	char *hexhash = tal_hexstr(tmpctx, hash, sizeof(struct sha256));
+	if (!shachain_add_hash(&chain->chain, index, hash))
+		return false;
+	if (!db_exec(__func__, wallet->db,
+		     "UPDATE shachains SET num_valid=%d, min_index=%" PRIu64
+		     " WHERE id=%" PRIu64,
+		     chain->chain.num_valid, index,
+		     chain->id)) {
+		goto fail;
+	}
+
+	if (!db_exec(__func__, wallet->db, "REPLACE INTO shachain_known "
+					   "(shachain_id, pos, idx, hash) VALUES "
+					   "(%"PRIu64", %d, %"PRIu64", '%s');",
+		     chain->id, pos, index, hexhash)) {
+		goto fail;
+	}
```

Comment:
True, but this is called from `wallet_channel_save` and is wrapped in a transaction there, but you're right, being able to call this outside of a transaction is bad. The problem is that sqlite3 doesn't support nested transactions. We can either `assert` that we are in a transaction or we can simulate nested transactions by incrementing a counter in `db_begin_transaction` and decrementing on `db_commit_transaction` and only really commit on 0. What do you think is better?